[ch124] Simple Dataset Details UI

seerocode opened the pull request:
Story details: https://app.clubhouse.io/stanford-computational-policy-lab/story/124

![image](https://user-images.githubusercontent.com/10480484/113452501-1cc4b200-93d2-11eb-81af-fe84aa908415.png)

Commit:
feat(ch124): add route for dataset details view

Commit:
feat(ch124): set up dataset details component

Commit:
refactor(ch124): refactor home, header, sidebar
- home comp: wrap a router link around the 'add data' button for each dataset
- header: add a link to home for the logo
- sidebar: wrap a router link around each dataset in sidebar

Commit:
style(ch124): fix linting

Commit:
test(ch124): fix failed snapshot test

Commit:
feat(ch124): add component for dataset records

Commit:
feat(ch124): update dataset details ui with mock data

seerocode commented on client/src/components/DatasetRecordsTable.tsx:
```diff
@@ -0,0 +1,209 @@
+        <Column
+          title="Black"
+          dataIndex="race"
+          render={(race: Race) => {
```
There's some duplication with these rendered functions and I'm working on refactoring that.

jnu replied:
some duplication is ok -- it's not usually worth generalizing on the first pass -- but it's a good instinct and i'll see if i can make some suggestions!

seerocode commented:
Adding tests also!

jnu commented on client/src/components/DatasetRecordsTable.tsx:
```diff
@@ -0,0 +1,209 @@
+          title="Black"
+          dataIndex="race"
+          render={(race: Race) => {
+            if (typeof race && typeof race.black !== "undefined") {
```
you've opened a big can o' worms in javascript ... _undefined_

`typeof race` always evaluates to true. if `race` was never defined, `typeof race === 'undefined'`. since `'undefined'` is a non-empty string, it will be interpreted as `true`.

`typeof race.black` might be `'undefined'`, but there are more idiomatic ways to check this. i'd suggest rewriting this function as something like:

```ts
(race: Race) => {
  if (!race || !race.hasOwnProperty('black') {
    return 0;
  }
  return race.black;
}
```

there's also some slight funniness with your function signature -- if you expect `race` might not exist, it should be marked as optional in the argument type. under what circumstances is `Race` undefined?

jnu commented on client/src/components/DatasetRecordsTable.tsx:
```diff
@@ -0,0 +1,209 @@
+          title="White"
+          dataIndex="race"
+          render={(race: Race) => {
+            if (typeof race && typeof race.white !== "undefined") {
```
the pattern i'd use to generalize this is a factory function. this uses a closure to create a function that comes configured to use a given attribute:

```ts
const getRaceRenderer = (raceKey: keyof Race) => {
  return (race: Race?) => {
    if (!race) {
      return 0;
    }
    return race[raceKey] || 0;
  };
}
```

then down here you can use:
```ts
render={getRaceRenderer('white')}
```

let me know if you have questions or want to pair on this!! the `keyof` in particular might raise an eyebrow :)

seerocode replied:
@jnu thank you!!!! I wasn't sure how to pass a generic type and `keyof` is what I was looking for.

EDIT: I read this a little wrong. I had something similar to this but I wanted to know how to pass a generic Type that could be `Race` or `Ethnicity`, etc.

jnu commented on client/src/components/DatasetRecordsTable.tsx:
```diff
@@ -0,0 +1,209 @@
+  asian?: number;
+};
+
+// TODO: populate type based on dataset options for ethnicity
```
hehe, this lil `TODO` might turn into a big headache :)

let's circle back to this later, though -- i think this is a great start for now

jnu commented on client/src/components/DatasetRecordsTable.tsx:
```diff
@@ -0,0 +1,209 @@
+  disability?: Disability;
+}
+
+// TODO: add gql query to get dataset id records based on user
```
is it worth merging your mock server pull request and setting up the GQL query now, in this pull request?

jnu commented on client/src/components/DatasetRecordsTable.tsx:
```diff
@@ -0,0 +1,209 @@
+          }}
+        />
+        <Column
+          title="Western European"
```
as you mentioned in your TODO above, in the future we'll want to generate these columns dynamically, but this is fine for now!

jnu reviewed (approved):
this looks great to me! i made a couple of small comments about refactoring and types, but im going to pre-emptively approve to unblock you. let me know if you want to go over anything!

Commit:
refactor(ch124): clean up some duplication
- extract render function for record category keys (see #27 comments)
- add link to dataset details page on home page

Commit:
test(ch124): fix code for failing test
